homeassistant/components/pyscript/__init__.py

```diff
@@ -0,0 +1,293 @@
+            state_func=state_func,
+            handler_func=handler_func,
+            trig_time_func=trig_time_func,
+        )
+        for trig in triggers.values():
+            trig.start()
+
+    hass.services.async_register(DOMAIN, SERVICE_RELOAD, reload_scripts_handler)
+
+    async def state_changed(event):
+        var_name = event.data["entity_id"]
+        # attr = event.data["new_state"].attributes
+        new_val = event.data["new_state"].state
+        old_val = event.data["old_state"].state if event.data["old_state"] else None
+        new_vars = {var_name: new_val, f"{var_name}.old": old_val}
+        func_args = {
+            "trigger_type": "state",
+            "var_name": var_name,
+            "value": new_val,
+            "old_value": old_val,
+        }
+        await state_func.update(new_vars, func_args)
+
+    async def start_triggers(event):
+        _LOGGER.debug("adding state changed listener")
+        hass.bus.async_listen(EVENT_STATE_CHANGED, state_changed)
+        _LOGGER.debug("starting triggers")
+        for trig in triggers.values():
+            trig.start()
+
+    async def stop_triggers(event):
+        _LOGGER.debug("stopping triggers")
+        for trig in triggers.values():
+            await trig.stop()
+
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STARTED, start_triggers)
+    hass.bus.async_listen(EVENT_HOMEASSISTANT_STOP, stop_triggers)
+
+    return True
+
+
+@bind_hass
+async def compile_scripts(
+    hass, event_func=None, state_func=None, handler_func=None, trig_time_func=None
+):
+    """Compile all python scripts in FOLDER."""
+
+    path = hass.config.path(FOLDER)
+
+    _LOGGER.debug("compile_scripts: path = %s", path)
+
+    def pyscript_service_factory(name, func, sym_table):
+        async def pyscript_service_handler(call):
+            """Handle python script service calls."""
+            # ignore call.service
+            _LOGGER.debug("service call to %s", name)
+            #
+            # use a new AstEval context so it can run fully independently
+            # of other instances (except for global_sym_table which is common)
+            #
+            ast_ctx = AstEval(
+                name,
+                global_sym_table=sym_table,
+                state_func=state_func,
+                event_func=event_func,
+                handler_func=handler_func,
+            )
+            handler_func.install_ast_funcs(ast_ctx)
+            func_args = {
+                "trigger_type": "service",
+            }
+            func_args = func_args.update(call.data)
+            handler_func.create_task(func.call(ast_ctx, [], call.data))
+
+        return pyscript_service_handler
+
+    triggers = {}
+    services = set()
+    for file in glob.iglob(os.path.join(path, "*.py")):
+        name = os.path.splitext(os.path.basename(file))[0]
+        _LOGGER.debug("parsing %s", file)
+        with open(file) as file_desc:
+            source = file_desc.read()
```

Comment:
This is doing I/O inside the event loop and should be moved to executor.

Reply:
Good catch - rookie error.  Moved all file operations to executor_job in commit below.